dapp/src/node/mod.rs

```diff
@@ -0,0 +1,378 @@
+
+        // Create a shared node instance
+        let node = Arc::new(RwLock::new(Node {
+            evm,
+            config,
+            blockchain: Blockchain::default(),
+            _s: PhantomData,
+        }));
+
+        // If node is configured to automine blocks, spawn a new thread to periodically mine blocks
+        if let Some(block_time) = automine {
+            let _shared_node = node.clone();
+            tokio::spawn(async move {
+                let mut interval = tokio::time::interval(block_time.into());
+                loop {
+                    interval.tick().await;
+                    // TODO: mine a new block
+                }
+            });
+        }
+
+        // Create a service with the shared node's state and serve it
+        let svc = Router::new()
+            .route("/", post(handler::<E, S>))
+            .layer(AddExtensionLayer::new(node))
+            .into_make_service();
+        let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
+        Server::bind(&addr).serve(svc).await.unwrap();
+    }
+}
+
+impl<S, E> Node<S, E>
+where
+    E: Evm<S>,
+{
+    /// Gets the native token balance of the account
+    pub fn get_balance(&self, account: Address) -> U256 {
+        self.evm.get_balance(account)
+    }
+
+    /// Gets the transaction by txhash
+    pub fn get_transaction(&self, tx_hash: TxHash) -> Option<Transaction> {
+        self.blockchain.tx(tx_hash)
+    }
+
+    /// Gets the transaction receipt by txhash
+    pub fn get_tx_receipt(&self, tx_hash: TxHash) -> Option<TransactionReceipt> {
+        self.blockchain.tx_receipt(tx_hash)
+    }
+
+    /// Gets the block by its number in the blockchain
+    pub fn get_block_by_number(&self, n: U64) -> Option<Block<TxHash>> {
+        self.blockchain.block_by_number(n)
+    }
+
+    /// Gets the block by its hash
+    pub fn get_block_by_hash(&self, hash: H256) -> Option<Block<TxHash>> {
+        self.blockchain.block_by_hash(hash)
+    }
+
+    fn accounts(&self) -> Vec<Wallet<SigningKey>> {
+        self.config.accounts.values().cloned().collect()
+    }
+
+    fn account(&self, address: Address) -> Option<Wallet<SigningKey>> {
+        self.config.accounts.get(&address).cloned()
+    }
+
+    fn default_sender(&self) -> Wallet<SigningKey> {
+        self.config
+            .accounts
+            .values()
+            .into_iter()
+            .next()
+            .cloned()
+            .expect("node should have at least one account")
+    }
+}
+
+impl<S, E> Node<S, E>
+where
+    E: Evm<S>,
+{
+    /// Simulates sending a transaction on the blockchain and returns the txhash
+    pub fn send_transaction(&mut self, tx: TypedTransaction) -> Result<TxHash, BoxedError> {
+        let sender = if let Some(from) = tx.from() {
+            if let Some(sender) = self.account(*from) {
+                sender
+            } else {
+                return Err(Box::new("account has not been initialized on the node"));
```

Comment:
Is the boxed error a requirement here? Can we instead define a `thiserror::Error` for all of these errors?